File: src/com/eichinn/composite/universal/Composite.java
package com.eichinn.composite.universal;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by ei_chinn on 2017/5/31.
 */
public class Composite extends Component {
    private List<Component> childComponents = null;

    @Override
    public void someOperation() {
        if (childComponents != null) {
            for (Component childComponent : childComponents) {
                childComponent.someOperation();
            }
        }
    }

    @Override
    public void addChild(Component child) {
        if (childComponents == null) {
            childComponents = new ArrayList<>();
        }
        childComponents.add(child);
    }

    @Override
    public void removeChild(Component child) {
        if (childComponents != null) {
            childComponents.remove(child);
        }
    }

    @Override
    public Component getChildren(int index) {
        if (childComponents != null && index >= 0 && index < childComponents.size()) {
            return childComponents.get(index);
        }
        return null;
    }
}

File: src/com/eichinn/visitor/extension/Composite.kt
package com.eichinn.visitor.extension

import java.util.*

/**
 * Created by ei_chinn on 2017/6/7.
 */
class Composite(name: String) : Component(name) {

    override fun accept(visitor: Visitor) {
        visitor.visitComposite(this)
        for (child in childComponents) {
            child.accept(visitor)
        }
    }

    override fun someOperation() {
        println("some operation in composite")
    }

    private val childComponents = ArrayList<Component>()

    override fun addChild(child: Component) {
        childComponents.add(child)
    }

    override fun removeChild(child: Component) {
        childComponents.remove(child)
    }

    override fun getChildren(index: Int): Component {
        return childComponents.get(index)
    }
}